tools/infrastructure/api_compare.py

```diff
@@ -0,0 +1,368 @@
+        if attr_name in mob_param_attributes:
+            mobile_attribute_value = mob_param_attributes[attr_name]
+        if attr_name in hmi_param_attributes:
+            hmi_attribute_value = hmi_param_attributes[attr_name]
+        values_equal = compare_func(
+            mobile_attribute_value, hmi_attribute_value)
+        if not values_equal:
+            hmi_result[attr_name] = hmi_attribute_value
+            mobile_result[attr_name] = mobile_attribute_value
+    return mobile_result, hmi_result
+
+
+global_compare_rules = [
+    (lambda mob_param_attributes, hmi_param_attributes:
+        mob_param_attributes["type"] == "String", StringCompareRule),
+    (lambda mob_param_attributes, hmi_param_attributes:
+        'array' in mob_param_attributes.keys() +
+        hmi_param_attributes.keys(), ArrayCompareRules),
+    (lambda mob_param_attributes, hmi_param_attributes:
+        True, AllCompareRules)
+]
+
+
+attributes_comparison_rules = {
+    'mandatory': lambda mobile_attribute_value, hmi_attribute_value:
+    True if mobile_attribute_value is None and hmi_attribute_value == "true"
+    else mobile_attribute_value == hmi_attribute_value,
+    'type': lambda mobile_attribute_value, hmi_attribute_value:
+    True if ("Common." + mobile_attribute_value) == hmi_attribute_value
+    else mobile_attribute_value == hmi_attribute_value
+}
+
+
+def simple_values_comparison(mobile_attribute_value, hmi_attribute_value):
+    return mobile_attribute_value == hmi_attribute_value
+
+
+# Function which converts json to xml format
+def json2xml(json_obj, line_padding=" "):
+    result_list = list()
+
+    json_obj_type = type(json_obj)
+
+    if json_obj_type is list:
+        for sub_elem in json_obj:
+            result_list.append(json2xml(sub_elem, line_padding))
+
+        return "\n".join(result_list)
+
+    if json_obj_type is dict:
+        for tag_name in json_obj:
+            sub_obj = json_obj[tag_name]
+            result_list.append("%s<%s>" % (line_padding, tag_name))
+            result_list.append(json2xml(sub_obj, "\t" + line_padding))
+            result_list.append("%s</%s>" % (line_padding, tag_name))
+
+        return "\n".join(result_list)
+
+    return "%s%s" % (line_padding, json_obj)
+
+
+def GetRpcParams(function):
+    params = {}
+    for function_tag in function:  # loop over function tags
+        if function_tag.tag == 'param':
+            name = function_tag.attrib['name']
+            params[name] = function_tag.attrib
+    return params
+
+
+def ExtractFunctions(interface, functions):
+    for child in interface:  # loop over every tag in interface
+        if child.tag == 'function':  # handle only when mobile tag is function
+            rpc_name = child.attrib['name']
+            rpc_type = child.attrib['messagetype']
+            key = Rpc(rpc_name=rpc_name, rpc_type=rpc_type)
+            params = GetRpcParams(child)
+            if key in functions:
+                functions[key].update(params)
```

Comment:
I recommend to check `function` isinstance equality to dictioanry at the beginning of method